vnext/Microsoft.ReactNative.Cxx/JSI/JsiApi.h

```diff
@@ -7,8 +7,55 @@
 #include "../ReactContext.h"
 #include "JsiAbiApi.h"
 
+// Use __ImageBase to get current DLL handle.
+// http://blogs.msdn.com/oldnewthing/archive/2004/10/25/247180.aspx
+extern "C" IMAGE_DOS_HEADER __ImageBase;
+
 namespace winrt::Microsoft::ReactNative {
 
+// Get JSI Runtime from the current JS dispatcher thread.
+// If it is not found, then create it and store it in the context.Properties().
+// Make sure that the JSI runtime holder is removed when the instance is unloaded.
+facebook::jsi::Runtime &GetOrCreateContextRuntime(ReactContext const &context) noexcept {
+  ReactDispatcher jsDispatcher = context.JSDispatcher();
+  VerifyElseCrashSz(jsDispatcher.HasThreadAccess(), "Must be in JS thread");
+
+  // The JSI runtime is not available if we do Web debugging when JS is running in web browser.
+  JsiRuntime abiJsiRuntime = context.Handle().JSRuntime().as<JsiRuntime>();
+  VerifyElseCrashSz(abiJsiRuntime, "JSI runtime is not available");
+
+  // See if the JSI runtime was previously created.
+  JsiAbiRuntime *runtime = JsiAbiRuntime::GetFromJsiRuntime(abiJsiRuntime);
+  if (!runtime) {
```

Comment:
>runtime [](start = 7, length = 7)

Do we need a mutex here to avoid multiple initialization?

Reply:
No, we do not need a mutex: it is all must happen in JS queue. At the begging of the function we crash if this assumption fails.

---
In reply to: [542951899](https://github.com/microsoft/react-native-windows/pull/6745#discussion_r542951899) [](ancestors = 542951899)